### PR TITLE
Refactor parameters as a JSON Object

### PR DESCRIPTION
**Pull request by csgt**
Refactor parameters as a JSON Object, add clientName and clientShortName parameters.  Adds flexibility for future customizations.  Update README to reflect the changes.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -496,9 +502,7 @@ export default class WhatsApp {
   private qrCodeListeners: (() => void)[] = [];
 
   constructor(
-    qrPath = "./qrcode.png",
-    restoreSession = false,
-    keysPath = "./keys.json"
+    parameters = {}
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
    opts = {      
        keysPath: "./keys.json",
        qrPath: "./qrcode.png",
        clientName: "WhatsApp forwarder",
        clientShortName: "WhatsAppForwarder",
        restoreSession: false 
    }
```
Rewriting all the options later is unnecessary.

File: src/main.ts [hunk 2]
```diff
@@ -508,17 +512,25 @@ export default class WhatsApp {
...
     this.keyPair = null;
 
-    this.qrPath = resolvePath(".", qrPath);
+    this.parameters = {...{
+      keysPath : "./keys.json",
+      qrPath : "./qrcode.png",
+      clientName : "WhatsApp forwarder",
+      clientShortName : "WhatsAppForwarder",
+      restoreSession : false
+    }, ...parameters}
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
    this.opts = Object.assign({}, opts, this.opts);
```
It's more readable in my opinion and performs better (according to [this site](https://jsperf.com/comparing-object-asign-spread))

File: src/main.ts [hunk 3]
```diff
@@ -508,17 +512,25 @@ export default class WhatsApp {
...
+
+    this.parameters.qrPath = resolvePath(".", this.parameters.qrPath);
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
    this.opts.qrPath = resolvePath(".", this.opts.qrPath);
```
Update the property name

File: src/main.ts [hunk 4]
```diff
@@ -508,17 +512,25 @@ export default class WhatsApp {
...
 
-    if (restoreSession) {
-      this.keysPath = resolvePath(".", keysPath);
+    if (this.parameters.restoreSession) {
+      this.parameters.keysPath = resolvePath(".", this.parameters.keysPath);
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
    if (this.opts.restoreSession) {
      this.opts.keysPath = resolvePath(".", this.opts.keysPath);
```
Update again

File: src/main.ts [hunk 5]
```diff
@@ -508,17 +512,25 @@ export default class WhatsApp {
...
     }
 
-    this.apiSocket.onopen = this.init(loginMsgId, restoreSession);
+    this.apiSocket.onopen = this.init(loginMsgId, this.parameters.restoreSession);
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
    this.apiSocket.onopen = this.init(loginMsgId, this.opts.restoreSession);
```
Update again

File: src/main.ts [hunk 6]
```diff
@@ -508,17 +512,25 @@ export default class WhatsApp {
...
     this.loginMsgId = loginMsgId;
 
-    if (restoreSession) {
-      doesFileExist(this.keysPath!).then(doesExist => {
+    if (this.parameters.restoreSession) {
+      doesFileExist(this.parameters.keysPath!).then(doesExist => {
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
    if (this.opts.restoreSession) {
      doesFileExist(this.opts.keysPath!).then(doesExist => {
```
Update property again

File: src/main.ts [hunk 7]
```diff
@@ -573,7 +585,7 @@ export default class WhatsApp {
 
   private saveKeys() {
     writeFile(
-      this.keysPath!,
+      this.parameters.keysPath!,
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
      this.opts.keysPath!,
```
Update property again

File: src/main.ts [hunk 8]
```diff
@@ -587,7 +599,7 @@ export default class WhatsApp {
 
   private async getKeys() {
     return new Promise((resolve, reject) => {
-      readFile(this.keysPath!, "utf-8", (err, data) => {
+      readFile(this.parameters.keysPath!, "utf-8", (err, data) => {
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
      readFile(this.opts.keysPath!, "utf-8", (err, data) => {
```
Update property again

File: src/main.ts [hunk 9]
```diff
@@ -692,7 +704,7 @@ export default class WhatsApp {
             this.setupEncryptionKeys(data as WhatsAppConnPayload);
             setTimeout(this.keepAlive.bind(this), 20 * 1000);
 
-            if (this.keysPath) {
+            if (this.parameters.keysPath) {
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
            if (this.opts.keysPath) {
```
Update property again

File: src/main.ts [hunk 10]
```diff
@@ -712,7 +724,7 @@ export default class WhatsApp {
 
             setTimeout(this.keepAlive.bind(this), 20 * 1000);
 
-            if (this.keysPath) {
+            if (this.parameters.keysPath) {
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
            if (this.opts.keysPath) {
```
Update property again

File: src/main.ts [hunk 11]
```diff
@@ -1585,7 +1597,7 @@ export default class WhatsApp {
       await qrcode.toDataURL(`${data.ref},${publicKeyBase64},${this.clientId}`)
     );
 
-    writeFile(this.qrPath, qrCode.data, err => {
+    writeFile(this.parameters.qrPath, qrCode.data, err => {
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
    writeFile(this.opts.qrPath, qrCode.data, err => {
```
Update property again

File: src/main.ts [hunk 12]
```diff
@@ -1595,18 +1607,18 @@ export default class WhatsApp {
     return async (e: WebSocket.OpenEvent) => {
       if (
         !restoreSession ||
-        (restoreSession && !(await doesFileExist(this.keysPath!)))
+        (restoreSession && !(await doesFileExist(this.parameters.keysPath!)))
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
        (restoreSession && !(await doesFileExist(this.opts.keysPath!)))
```
Update property again

File: src/main.ts [hunk 13]
```diff
@@ -1595,18 +1607,18 @@ export default class WhatsApp {
...
       ) {
         this.clientId = crypto.randomBytes(16).toString("base64");
       } else {
         await this.getKeys();
       }
 
       e.target.send(
-        `${loginMsgId},["admin","init",[0,4,2080],["WhatsApp forwarder","WhatsAppForwarder"],"${this.clientId}",true]`
+        `${loginMsgId},["admin","init",[0,4,2080],["${this.parameters.clientName}","${this.parameters.clientShortName}"],"${this.clientId}",true]`
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
        `${loginMsgId},["admin","init",[0,4,2080],["${this.opts.clientName}","${this.opts.clientShortName}"],"${this.clientId}",true]`
```
Update property again

File: src/main.ts [hunk 14]
```diff
@@ -1595,18 +1607,18 @@ export default class WhatsApp {
...
       );
 
-      if (restoreSession && (await doesFileExist(this.keysPath!))) {
+      if (restoreSession && (await doesFileExist(this.parameters.keysPath!))) {
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
      if (restoreSession && (await doesFileExist(this.opts.keysPath!))) {
```
Update property again

File: src/main.ts [hunk 15]
```diff
@@ -458,8 +458,14 @@ export interface WAStubMessage {
...
 
-  private keysPath?: string;
-  private qrPath: string;
+  private parameters : {
+    keysPath: string,
+    qrPath: string,
+    clientName : string,
+    clientShortName : string,
+    restoreSession : boolean
+  }
```
Comment (noamalffasy) [inline src/main.ts]:
```suggestion
  private opts: {
    keysPath: string,
    qrPath: string,
    clientName: string,
    clientShortName: string,
    restoreSession: boolean
  }
```
Adjusting it to my preference and code style.

File: README.md [hunk 16]
```diff
@@ -20,10 +20,19 @@ yarn add @noamalffasy/js-whatsapp
...
-const wap = new Whatsapp();
+const params = {
+  keysPath : "./keys.json",
+  qrPath : "./qrcode.png",
+  clientName : "WhatsApp forwarder",
+  clientShortName : "WhatsAppForwarder",
+  restoreSession : false
+}
+
+const wap = new Whatsapp(params);
```
Comment (noamalffasy) [inline README.md]:
```suggestion
const opts = {
  keysPath: "./keys.json",
  qrPath: "./qrcode.png",
  clientName: "WhatsApp forwarder",
  clientShortName: "WhatsAppForwarder",
  restoreSession: false
}

const wap = new Whatsapp(opts);
```
I'd change the name to this instead, just a preference.
I also changed the code style (no space before the` :` and after the property name).

File: README.md [hunk 17]
```diff
@@ -41,7 +50,7 @@ const wap = new Whatsapp();
 to:
 
 ```js
-const wap = new Whatsapp(true);
+const wap = new Whatsapp({restoreSession : true});
```
Comment (noamalffasy) [inline README.md]:
```suggestion
const wap = new Whatsapp({restoreSession: true});
```
Same code style

**Review by noamalffasy — CHANGES_REQUESTED** [review]
That's a great suggestion, I'd change a few things (I added suggestions to them, you should be able to just agree to them and automatically update it).

In addition to them, I think that along with this change I should also update all of the methods and make them use JSON objects too.

I wanted to keep it for a later version so it wouldn't break existing code. However, since I'm bringing this one over now, I think that updating the functions should be done along with it.

What do you think of the suggestion?

**Comment by csgt** [comment]
Yeah, opts looks fine.  Maybe tag a major version change so current users are not affected and apply the params to all functions!

**Pull request closed without merging**